lib/dao.js

```diff
@@ -2689,6 +2689,12 @@ DataAccessObject.replaceById = function(id, data, options, cb) {
   Model.notifyObserversOf('before save', context, function(err, ctx) {
     if (err) return cb(err);
 
+    if (ctx.instance[pkName] !== id && !(Model._warned || {}).cannotOverwritePKInBeforeSaveHook) {
+      Model._warned = { cannotOverwritePKInBeforeSaveHook: true };
```

Comment:
When we add more warning in the future, this line will clear all other warnings that may have been already reported/set.

I am proposing to initialize `Model._warned = {}`in the code building Model ctor, so that we can rely on it being initialized everywhere else. See https://github.com/strongloop/loopback-datasource-juggler/blob/b039b51610b5ad14b5abdc5bb9e798baa7c70588/lib/model-builder.js#L222-L233. I would use `hiddenProperty` to define `_warned`, so that we are consistent with other ctor-level properties.